### PR TITLE
introduced logging, SDK version control

### PR DESCRIPTION
**Pull request by martinezpl**
In the future, we can use the custom logging formate to send analytics reports about SDK usage, errors etc. :) 

I tested this by adding Warning header in the source code - tested this way everything works as intended. Cannot write a test for this until we return Warning headers from our API.

### CODE REVIEW

File: apiaudio/__init__.py [hunk 1]
```diff
@@ -18,8 +17,9 @@
...
 
-# Set to either 'debug' or 'info', controls console logging
-log = None
-
 # API resources
 from apiaudio.api_resources import *
+
+# logging
+from apiaudio.logging import SDKLogger
+APILogger = SDKLogger("apiaudio")
```
Comment (martinezpl) [inline apiaudio/__init__.py:25]: I'm thinking one call during initiation? A specific endpoint that we can use to log user activity and/or return whatever updates we want?

File: apiaudio/api_request.py [hunk 2]
```diff
@@ -141,8 +143,10 @@ def _expanded_raise_for_status(self, res):
         @return: None
         """
         try:
-            if res.headers.get("Warning"):
-                print(res.json().get("warnings"))
+            if res.headers.get("Warning", "e"):
+                for warn in res.json().get("warnings", ["e"]):
```
Comment (zeritte) [inline apiaudio/api_request.py:147]: ['e']

File: apiaudio/__init__.py (hunk 1 above)
Comment (Sjhunt93) [reply apiaudio/__init__.py:25]: I think we might have to be careful about sending back any information collected on the client-side (GDPR and whatnot)

**Review by Sjhunt93 — COMMENTED** [review]
I think in general this looks ok.

Can you prepare an end-to-end test for this? 

I.e. 1 test that will produce a warning response. If you branch from end-to-end staging branch and then send me a PR it will get picked up when we commit this branch into version flag changes also

**Comment by Sjhunt93** [comment]
Also.

Can you add a disable warnings message? So customers can also choice to turn this off

File: apiaudio/api_request.py [hunk 3]
```diff
@@ -141,6 +145,16 @@ def _expanded_raise_for_status(self, res):
         @return: None
         """
         try:
+            if res.headers.get("Warning") and apiaudio.log_warnings:
+                for warn in re.findall(r'\"(.*?)\"', res.headers["Warning"]):  # get all messages in between ""
+                    if apiaudio.sdk_version in warn:
```
Comment (zeritte) [inline apiaudio/api_request.py:150]: i did not understand this at all, why would `0.1.2` be in warning messages

**Review by Sjhunt93 — COMMENTED** [review]
In general, I think the Logger needs to be designed as a singleton instance. Right now every class has a logger, but there is no difference. Instead, you could pass object name to raise_for_status, and then pass this to a 'log_warning' function?


Otherwise we are creating a seperate logger instance for each class

**Comment by Sjhunt93** [comment]
> In general, I think the Logger needs to be designed as a singleton instance. Right now every class has a logger, but there is no difference. Instead, you could pass object name to raise_for_status, and then pass this to a 'log_warning' function?
> 
> Otherwise we are creating a seperate logger instance for each class

If you want to talk more about this we can

File: apiaudio/api_request.py (hunk 3 above)
Comment (martinezpl) [reply apiaudio/api_request.py:150]: If the warning is concerning the version being non-latest, the current version will be mentioned. I'm changing this as discussed in https://github.com/aflorithmic/api-utils/pull/12 this kind of warning will be signified by a separate header from tracing

File: apiaudio/api_request.py [hunk 4]
```diff
@@ -114,18 +117,20 @@ def _get_request(cls, url=None, path_param=None, request_params=None):
...
         local_filename = local_filename.replace("%243ct10n", "section")
         local_filename = local_filename.replace("%7C", "|")
 
-        with requests.get(url, stream=True) as r:
-            cls._expanded_raise_for_status(r)
+        response = requests.get(url, stream=True)
+        while response.status_code == 202:  # keep the retrieval going if 202 returned
+            response = requests.get(url, stream=True)
+
+        cls._expanded_raise_for_status(response)
```
Comment (martinezpl) [inline apiaudio/api_request.py:130]: here's the download fix to keep retrieving when 202 (in progress) is returned

**Review by zeritte — APPROVED** [review]

File: apiaudio/api_request.py (hunk 4 above)
Comment (Sjhunt93) [reply apiaudio/api_request.py:130]: Do we not need a timeout here? Like a maximum retries?

File: apiaudio/helper_classes.py [hunk 5]
```diff
@@ -17,32 +17,48 @@ class CreatableResource(APIRequest):
...
 class DeletableResource(APIRequest):
     @classmethod
     def delete(cls, **args):
 
-        if cls.OBJECT_NAME == 'media':
-            if 'mediaId' not in args:
-                raise Exception('mediaId parameter is required')
-            
-            return cls._delete_request(path_param=cls.audio_resource_path, request_params=args)
+        if cls.OBJECT_NAME == "media":
```
Comment (Sjhunt93) [inline apiaudio/helper_classes.py:25]: Im guessing this is all coming from black/linter? 

 Im guessing as this is the first time its run it probably should be a separate PR otherwise it makes it a little harder to review :D

**Review by Sjhunt93 — APPROVED** [review]
Approved but can you just reply to the two comments :D

File: apiaudio/helper_classes.py (hunk 5 above)
Comment (martinezpl) [reply apiaudio/helper_classes.py:25]: well, it's messy just once, from now on the whole source code is compliant with the standard :)

File: apiaudio/api_request.py (hunk 4 above)
Comment (martinezpl) [reply apiaudio/api_request.py:130]: I don't think so
